src/font_file_lister_gdi.cpp

```diff
@@ -16,284 +16,195 @@
+		throw agi::EnvironmentError("Failed to initialize the DirectWrite Factory");
+	dwrite_factory_sh = dwrite_factory;
+
+	IDWriteFontCollection* font_collection;
+	if (FAILED(dwrite_factory_sh->GetSystemFontCollection(&font_collection, true)))
+		throw agi::EnvironmentError("Failed to initialize the system font collection");
+	font_collection_sh = font_collection;
+
+	HDC dc = CreateCompatibleDC(nullptr);
+	if (dc == nullptr)
+		throw agi::EnvironmentError("Failed to initialize the HDC");
+	dc_sh = dc;
+
+	IDWriteGdiInterop* gdi_interop;
+	if (FAILED(dwrite_factory_sh->GetGdiInterop(&gdi_interop)))
+		throw agi::EnvironmentError("Failed to initialize the Gdi Interop");
+	gdi_interop_sh = gdi_interop;
 }
 
-std::vector<agi::fs::path> get_installed_fonts() {
-	std::vector<agi::fs::path> files;
+CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, int bold, bool italic, std::vector<int> const& characters) {
+	CollectionResult ret;
 
-	wchar_t fdir[MAX_PATH];
-	SHGetFolderPathW(NULL, CSIDL_FONTS, NULL, 0, fdir);
-	agi::fs::path font_dir(fdir);
+	int weight = bold == 0 ? 400 :
+				 bold == 1 ? 700 :
+				 bold;
 
-	// System fonts
-	read_fonts_from_key(HKEY_LOCAL_MACHINE, font_dir, files);
+	// From VSFilter
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/RTS.cpp#l45
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l2992
+	LOGFONTW lf{};
+	lf.lfCharSet = DEFAULT_CHARSET; // we should use the one specified in the ass file
+	wcsncpy_s(lf.lfFaceName, LF_FACESIZE, agi::charset::ConvertW(facename).c_str(), _TRUNCATE);
+	lf.lfItalic = italic ? -1 : 0;
+	lf.lfWeight = weight;
+	lf.lfOutPrecision = OUT_TT_PRECIS;
+	lf.lfClipPrecision = CLIP_DEFAULT_PRECIS;
+	lf.lfQuality = ANTIALIASED_QUALITY;
+	lf.lfPitchAndFamily = DEFAULT_PITCH | FF_DONTCARE;
+
+	agi::scoped_holder<HFONT> hfont_sh(CreateFontIndirect(&lf), [](HFONT p) { DeleteObject(p); });
+	if (hfont_sh == nullptr)
+		return ret;
 
-	// User fonts
-	read_fonts_from_key(HKEY_CURRENT_USER, font_dir, files);
+	SelectFont(dc_sh, hfont_sh);
 
-	return files;
-}
+	std::wstring selected_name(LF_FACESIZE - 1, L'\0');
+	if (!GetTextFaceW(dc_sh, LF_FACESIZE, &selected_name[0]))
+		return ret;
 
-using font_index = std::unordered_multimap<uint32_t, agi::fs::path>;
-
-font_index index_fonts(FontCollectorStatusCallback &cb) {
-	font_index hash_to_path;
-	auto fonts = get_installed_fonts();
-	std::unique_ptr<char[]> buffer(new char[1024]);
-	for (auto const& path : fonts) {
-		try {
-			auto stream = agi::io::Open(path, true);
-			stream->read(&buffer[0], 1024);
-			auto hash = murmur3(&buffer[0], stream->tellg());
-			hash_to_path.emplace(hash, path);
-		}
-		catch (agi::Exception const& e) {
-			cb(to_wx(e.GetMessage() + "\n"), 3);
+	// If the selected_name is different then the lf.lfFaceName,
+	// it means that the requested font doesn't exist.
+	if (_wcsnicmp(&selected_name[0], lf.lfFaceName, LF_FACESIZE))
```

Comment:
And then this could also be `selected_name.data()` for consistency (but this time it's just cosmetic).